### PR TITLE
Add SDRewardManager (ETHx Merkle Automation)

### PR DESCRIPTION
**Pull request by blockgroot**
SDRewardManager contract is responsible to transfer SD rewards to the socialise pool.
It has mechanism to add entry for the current cycle and approver can approve this entry.

### CODE REVIEW

**Comment by galacticminter** [comment]
@manoj9april Could you take a look at this please?

File: contracts/StaderConfig.sol
Comment (manoj9april) [inline contracts/StaderConfig.sol:307]: @blockgroot I believe `MANAGER` would not be able to grantRoles ? only `DEFAULT_ADMIN_ROLE` has permissions to grantRole ?

File: contracts/StaderConfig.sol
Comment (manoj9april) [inline contracts/StaderConfig.sol:561]: Lets add natspec comments as well, so that other devs are aware that this view function is useful only when called by specific contracts.
------
Shall we also expose one more view function with `contractAddress` as well ? this would helpful to manually check using etherscan ?

File: contracts/SDRewardManager.sol
Comment (manoj9april) [inline contracts/SDRewardManager.sol:60]: if allowed user happens to pass a cycle_number bigger than required cycle_number, sdRewardManager would be blocked.

I would suggest to allow user to send `cycleNumber` instead read the `cycleNumber` from `socialisingPool`/`staderOracle`. this will also result in getting rid of few extra `cycleNumber` checks.

similarly in approveEntry, only the current cycle should be read from socialisingPool

File: contracts/StaderConfig.sol
Comment (manoj9april) [inline contracts/StaderConfig.sol:301]: one concern about this approach is, if we upgrade the functionSig, we have to take care that we revoke permission from previous functionSig and grant permission to new functionSig. Hence this would have to be added as part of upgrade txns. Hence a little extra overhead. And hopefully people shouldn't miss this step.

File: test/foundry_tests/SDRewardsManager.t.sol
Comment (manoj9april) [inline test/foundry_tests/SDRewardsManager.t.sol:17]: typo : `SDRewardsManagerTest` 
missing `s`

File: contracts/StaderConfig.sol
Comment (manoj9april) [reply contracts/StaderConfig.sol:307]: ah, understood, you are using internal `_grantRole`

File: test/foundry_tests/SDRewardsManager.t.sol
Comment (blockgroot) [reply test/foundry_tests/SDRewardsManager.t.sol]: [49d6513](https://github.com/stader-labs/ethx/pull/253/commits/49d65130aaa1e19d52b13ddedc13e0a88212a0b8) given we have a single kind of reward, hence refactoring it to `SDRewardManager`

File: contracts/StaderConfig.sol
Comment (blockgroot) [reply contracts/StaderConfig.sol]: [27fc0bc](https://github.com/stader-labs/ethx/pull/253/commits/27fc0bc3dbf0b42bb598422518b4e73ea9702530) legacy mechanism used for access control

File: contracts/SDRewardManager.sol
Comment (blockgroot) [reply contracts/SDRewardManager.sol]: [837cf25](https://github.com/stader-labs/ethx/pull/253/commits/837cf25cc2c7604af369769a6963c61ffee61323)

File: contracts/SDRewardManager.sol
Comment (manoj9april) [inline contracts/SDRewardManager.sol:102]: shall we also add code for required sdToken approval to this contract ?

File: contracts/SDRewardManager.sol
Comment (manoj9april) [inline contracts/SDRewardManager.sol:87]: i believe we can omit this and read from rewardEntry ?

File: contracts/StaderConfig.sol
Comment (manoj9april) [inline contracts/StaderConfig.sol:70]: i believe we should also expose a write function to grant these Roles by MANAGER, as ADMIN is an external multisig and needn't be pulled incase we want to grant/revoke this role .
----
Else we can have ROLE_SD_REWARD_ENTRY_ACCOUNT an 1/n multisig, and approver also an multisig, so as there is little to no probability of changing thses addresses

**Review by manoj9april — COMMENTED** [review]
@galacticminter @blockgroot please check comments

File: contracts/StaderConfig.sol
Comment (blockgroot) [reply contracts/StaderConfig.sol:70]: To maintain traceability, I would keep it as it is. I agree if we have multisig for the new set of permissions. WDYT @galacticminter

File: contracts/SDRewardManager.sol
Comment (blockgroot) [reply contracts/SDRewardManager.sol:102]: It's a good idea but I think to maintain better traceability and avoid any human error while adding entry, It's a better idea to keep it decouple. WDYT?

Comment (manoj9april) [reply contracts/SDRewardManager.sol:102]: cool

**Review by galacticminter — APPROVED** [review]

**Pull request merged**
